### ISSUE TITLE
[css-transitions] smoothly transitioning scrollbars in/out when it appears/disappears

### ISSUE THREAD

**Issue by emattias** [w3c/csswg-drafts#6844]
When the scrollbar is shown/hidden can be very jarring and cause the page to jump. The scrollbar-gutter and scrollbar-width CSS properties gives us some control but to my knowledge it doesn't allow us to make the transition from scrollbar to no scrollbar (and vice versa) a transition rather than a jump.

If this is the wrong place to post this, please let me know where I should post it and I'll do it :)

**Comment by fantasai** [comment]
@emattias Are you concerned about the impact on layout being abrupt, or the visual appearance of the scrollbar being abrupt?

I think `scrollbar-gutter` was intended to solve the former. For the latter, seems to me this is more of a browser quality-of-implementation issue?

**Comment by emattias** [comment]
> I think `scrollbar-gutter` was intended to solve the former. For the latter, seems to me this is more of a browser quality-of-implementation issue?

Since everything else on the page can be smootly transitioned/animated I think this should also be able to do this. Similar to how you can do:

```css
html {
  scroll-behavior: smooth;
}
```

I think you should be able to do something like:
```css
html {
  scrollbar-transition: smooth;
}
```

(Any maybe with some options for timing etc.)

The problem with `scrollbar-gutter` is that it avoids the jump but instead you get an awkward space that covers content etc.

**Comment by bmathwig** [comment]
I agree with @fantasai in that this is a browser QoL issue. On Windows, we use a fade transition between 'shy' and 'visible' states of the scrollbar.

![Animation of Windows scrollbar hiding behavior!](https://docs.microsoft.com/en-us/windows/apps/design/controls/images/conscious-scroll.gif)

**Comment by emattias** [comment]
The problem is that `scrollbar-gutter` cant always be used because of the awkward space that gets left when the content is not overflowing (so no scrollbar). I dont think this space is stylable? See the white space to the right here:

<img width="133" alt="CleanShot 2022-06-27 at 22 39 36@2x" src="https://user-images.githubusercontent.com/351537/176031750-f8e3b0e2-e219-49c6-b668-c1a052a8fe47.png">

So now we have to have this abrupt jump when opening modals:

![CleanShot 2022-06-27 at 22 42 29](https://user-images.githubusercontent.com/351537/176033319-d083df73-4cda-4d6e-acce-c506d8ece506.gif)

Its especially noticable when a modal is animated in smoothly. Can't the browser provide a way to have this be smooth also?

Giving developers more control of the "empty scrollbar gutter" might be a way to solve this problem also.

**Comment by tabatkins** [comment]
I mean, that "space left when the content is not overflowing" is the entire point of scrollbar-gutter, yeah - it means you don't have *any* layout shift between the scrolling and non-scrolling space.

Can you give a markup example that's causing you problems with things not extending to the full width?

**Comment by bmathwig** [comment]
I noticed that you used the `scrollbar-gutter` on the `<header>` element itself. To control the gutter color, you can put a `<div>` inside of the `<header>` and apply `scrollbar-gutter` to the `<div>` rather than the `<header>`. This will make the internal contents of the `<div>` to be offset by the scrollbar but maintain the background color of the `<header>`.

**Comment by emattias** [comment]
@bmathwig I'm not following. scrollbar-gutter is applied to `<html>` in the codepen I linked to above. Could you make a prn showing your suggested change maybe?

Anyway, I still think scrollbar-gutter is unusable for many sites/apps because of the reasons I mentioned before.
